scipy/stats/_continuous_distns.py

```diff
@@ -5508,67 +5508,56 @@ def _entropy(self, s):
         this function uses explicit formulas for the maximum likelihood
         estimation of the log-normal shape and scale parameters, so the
         `optimizer`, `loc` and `scale` keyword arguments are ignored.
+        If the location is free, a likelihood maximum is found by
+        setting its partial derivative wrt to location to 0, and
+        solving by substituting the analytical expressions of shape
+        and scale (or provided parameters).
+        See, e.g., equation 3.1 in
+        A. Clifford Cohen & Betty Jones Whitten (1980)
+        Estimation in the Three-Parameter Lognormal Distribution,
+        Journal of the American Statistical Association, 75:370, 399-404
+        https://doi.org/10.2307/2287466
         \n\n""")
     def fit(self, data, *args, **kwds):
-        floc = kwds.get('floc', None)
-        if floc is None:
-            # fall back on the default fit method.
-            return super().fit(data, *args, **kwds)
-
-        f0 = (kwds.get('f0', None) or kwds.get('fs', None) or
-              kwds.get('fix_s', None))
-        fscale = kwds.get('fscale', None)
-
-        if len(args) > 1:
-            raise TypeError("Too many input arguments.")
-        for name in ['f0', 'fs', 'fix_s', 'floc', 'fscale', 'loc', 'scale',
-                     'optimizer', 'method']:
-            kwds.pop(name, None)
-        if kwds:
-            raise TypeError("Unknown arguments: %s." % kwds)
-
-        # Special case: loc is fixed.  Use the maximum likelihood formulas
-        # instead of the numerical solver.
-
-        if f0 is not None and fscale is not None:
-            # This check is for consistency with `rv_continuous.fit`.
-            raise ValueError("All parameters fixed. There is nothing to "
-                             "optimize.")
-
-        data = np.asarray(data)
+        parameters = _check_fit_input_parameters(self, data, args, kwds)
+        data, fshape, floc, fscale = parameters
 
-        if not np.isfinite(data).all():
-            raise ValueError("The data contains non-finite values.")
+        def get_shape_scale(loc):
+            if fshape is None or fscale is None:
+                lndata = np.log(data - loc)
+            scale = fscale or np.exp(lndata.mean())
+            shape = fshape or np.sqrt(np.mean((lndata - np.log(scale))**2))
+            return shape, scale
 
-        floc = float(floc)
-        if floc != 0:
-            # Shifting the data by floc. Don't do the subtraction in-place,
-            # because `data` might be a view of the input array.
-            data = data - floc
-        if np.any(data <= 0):
-            raise FitDataError("lognorm", lower=floc, upper=np.inf)
-        lndata = np.log(data)
+        def dL_dLoc(loc):
+            shape, scale = get_shape_scale(loc)
+            shifted = data - loc
+            return np.sum((1 + np.log(shifted/scale)/shape**2)/shifted)
```

Comment:
This appears correct, but I did a Monte Carlo experiment and found many instances in which this PR produces a fit inferior to that in main. Consider, e.g.
```python3
rng = np.random.default_rng(2433006235492611347)
s, loc, scale = (0.3600260879787981, -33.14629739153586, 1.1683787833422574)
size = 30
dist = stats.lognorm(s, loc, scale)
rvs = dist.rvs(size=size, random_state=rng)  #  (1.0257942280215272e-05, -50464.515839187385, 50432.702840596336)
s_fit, loc_fit, scale_fit = stats.lognorm.fit(rvs)  # 22.796260407433124
```
whereas in main:
```python3
rvs = dist.rvs(size=size, random_state=rng)  #  (0.30898245311983374, -33.40443838012905, 1.5165142163884346)
s_fit, loc_fit, scale_fit = stats.lognorm.fit(rvs)  # 19.826858766013423
```

I think the problem in this particular case is a numerical issue when `dL_dLoc` is evaluated at `rbrack` that causes the value of `dL_dLoc` to be positive, whereas it should be negative - and it is for ``dL_dLoc(np.min(data)-1e-12)`, but not when the argument gets really close to `np.min(data)`. 

Then, we seeks a value for `lbrack` that causes a sign change. It can't find one until it goes very far out toward negative infinity, where in reality the derivative is nearly zero, but it eventually finds a value that is slightly negative (probably due to numerical noise). It `root_scalar` then converges to a `loc` quite far out to the left, whereas the optimal location is actually quite close to `np.min(data)`.

I'd suggest trying to find an alternative formulation that avoids this issue when evaluated at the current `rbrack`. If not, you might need to search for an `rbrack` that avoids this issue.

This brings to mind another possible issue - is the optimal solution ever at `loc = rbrack`? That is, is it possible for the analytical value of the partial derivative of the log likehood with respect to `loc` to be positive when evaluated at `rbrack`? If so, you might need to treat this case separately, because the solution is not when the partial derivative is zero but instead it when the constraint `loc < np.min(data)` is active.

Reply:
That is brought up in the reference. The log-likelihood tends to `inf` for `shape, loc, scale = inf, min(data), 0+`. They hinted it is more common with small samples.
Using Newton's method (as mentioned [above](https://github.com/scipy/scipy/pull/16839#discussion_r946102104)), with a starting point close to `min(data)` but much further away (like a few units) could be a way to get around it?

Reply:
I tried it with numerical derivatives (`optimize.approx_fprime`) and it usually fails to converge. Maybe you'll have better luck. You could try providing an analytical derivative - but it's not so simple because you can't take the (second) derivative holding the optimal scale and shape fixed; you also need to consider the partial derivatives of the optimal scale and shape w.r.t. to the location.

Reply:
For this PR, unless you see an easy fix, why don't you detect when `dL_dLoc` is positive at `rbrack` and in that case, revert to super.

Reply:
Here is what I did.
I determined the analytical derivative to dL_dLoc (actually, `dL_dLoc` times the `shape` squared, as it is a bit nicer) so I can provide it to `root_scalar`. If `dL_dLoc(rbrack) >= 0`, or an initial guess has been provided, I use it. Otherwise, I keep to the bracketing method. `dL_dLoc` is quite unstable for negative large floats, so I still iteratively look for a bracket, bypassing the `lbeack > -np.inf` test and catching the `ValueError` should it be raised.
Running tests now, I will push when it is done, but if works nicely for your case (same result as main, 14 times faster on my laptop).

Reply:
Great. And I can post my Monte Carlo experiment code if you would like.

Reply:
Sure!
I added a call to `_assert_less_or_close_loglike` in my hypothesis file. Another set of issues was linked to `dL_dLoc(rbrack) < 0`, but becoming positive then negative again in a very small interval. So I ended up prioritizing the method involving the derivative, but it has shortcomings as well, that cause the warnings mentioned [here](https://github.com/scipy/scipy/pull/16839#discussion_r945912194) as the optimisation often fails. I realised the benches got a bit long, and it is because the method ends up falling back to `super`. I realised it after pushing, so I will give it another try tomorrow; iteratively bracketing from the MM estimate rather than using it as a start for Newton's method, something like
```
rbrack = (rbrack + min(data))/2
lbrack = 2*lbrack - min(data)
```
maybe.

Reply:
OK. When you're happy with it, please post the results of the experiment like [mine](https://github.com/scipy/scipy/pull/16839#issuecomment-1217437632) (code provided) and mention me.

Reply:
@mdhaber , here is where I'm at now.

The method of moment estimate tended to be a poor initial guess for Newton's method, so I dropped this method to focus on Brent's. Using the estimate as a starting point for determining the bracket solves the "`rbrack` too close to `min(data)`" problem.

Much less falling back to `super`, by looking at the execution time histogram.
![histo_time](https://user-images.githubusercontent.com/34381942/185624883-2c0d7c2b-661b-4d86-a256-cc911e0539be.png)

I do not quite get why, but the improvement histogram is very similar to yours, with only about 70% of improvements
(697 negative `nllf_diff`, 30 zeros, 178 positive less than 1, so 95 worse than that).
![histo_improvement](https://user-images.githubusercontent.com/34381942/185624879-cd00e83a-1364-4aca-9b54-481aba0903cb.png)

I guess these cases were covered by defaulting to `super` in the previous commit.
But I got hypothesis to pass all calls to `_assert_less_or_close_loglike` through the parameter space and for all combinations of fixed parameters (with the occasional flakiness), when assuming
- distinct elements in the array;
- a positive skewness.

I believe these are not too restrictive for real world data that is expected to be log-normally distributed. Only one of the positive `nllf_diff` is filtered out when applying these conditions, though. Not sure where does the mismatch come from.

Reply:
> 178 positive less than 1, so 95 worse than that.

These are relative errors, so "less than 1" can be a pretty substantial regression. This is is the plot:

![image](https://user-images.githubusercontent.com/6570539/185730390-d3ccea8a-d70f-412b-bad8-8e33c01258d5.png)

I'm not sure we'd want to do that.

---

https://github.com/scipy/scipy/pull/16839#discussion_r946216467:
> For this PR, unless you see an easy fix, why don't you detect when `dL_dLoc` is positive at `rbrack` and in that case, revert to super.

I tried this; I went back to the original commit, 3bcfcc5, and made one change - falling back to `super` if `dL_dLoc(rbrack) > 0`. 

Here's are the results:
![image](https://user-images.githubusercontent.com/6570539/185730096-1d9b4c20-4c9a-4f35-a1bc-5a6128f4c282.png)

![image](https://user-images.githubusercontent.com/6570539/185730121-ec114f50-743b-4d7a-b85f-f4e4f91ccb6b.png)

There are only five "regressions", which are inconsequential:
```
nllf_diff_bad  # array([1.16510943e-16, 1.60065708e-16, 1.82261732e-16, 1.32028172e-16,
       1.43490002e-16])
```

And 332 cases in which there is zero difference in NLLF.

Since that was a bit simpler and I already reviewed it, what do you think about just doing that? This would still be a big improvement in most cases.

Reply:
One last twist; I found [this paper](https://rss.onlinelibrary.wiley.com/doi/abs/10.2307/2346411) suggesting maximising the log-likelihood (as a function of the location) with a change of variable `eta = -log(min(data)-loc`. When stripping all the constants, the negative log-likelihood is just `log(scale) + log(shape)`, with both `scale` and `shape` functions of `loc` (or fixed).
The advantages being that there is no restriction on `eta`, which can span the real line (even though it numerical problems can still arise for small values); and no need to iteratively determine a bracketing interval for `minimize_scalar`.
I gave it a go and results are pretty similar to the previous "working" commit.

![histo_time](https://user-images.githubusercontent.com/34381942/185764026-ac40dd29-4031-4e41-9380-2b336b8a739b.png)
![histo_improvement](https://user-images.githubusercontent.com/34381942/185764024-cd5962ca-b968-4f0a-aa29-4aa84bde463a.png)
```
sum(times_pr < times_super)  # 735
sum(nllf_diff < 0), sum(nllf_diff == 0), sum(nllf_diff > 0)  # (638, 359, 3)
nllf_diff_bad  #  array([1.16510943e-16, 1.17851497e-16, 1.60065708e-16])
```

Reply:
OK. If that's about as simple (or simpler) than what I suggested yesterday (https://github.com/scipy/scipy/pull/16839#discussion_r950652435) go ahead and push it.